crates/noirc_frontend/src/node_interner.rs

```diff
@@ -585,16 +587,67 @@ impl NodeInterner {
         method_id: FuncId,
     ) -> Option<FuncId> {
         match self_type {
-            Type::Struct(struct_type, generics) => {
+            Type::Struct(struct_type, _generics) => {
                 let key = (struct_type.borrow().id, method_name);
-                self.struct_methods.insert(key, (generics.clone(), method_id)).map(|(_, id)| id)
+                self.struct_methods.insert(key, method_id)
+            }
+            Type::Error => None,
+
+            other => {
+                let key = get_type_method_key(self_type).unwrap_or_else(|| {
+                    unreachable!("Cannot add a method to the unsupported type '{}'", other)
+                });
+                self.primitive_methods.insert((key, method_name), method_id)
             }
-            other => unreachable!("Tried adding method to non-struct type '{}'", other),
         }
     }
 
     /// Search by name for a method on the given struct
     pub fn lookup_method(&self, id: StructId, method_name: &str) -> Option<FuncId> {
-        self.struct_methods.get(&(id, method_name.to_owned())).map(|(_, id)| *id)
+        self.struct_methods.get(&(id, method_name.to_owned())).copied()
+    }
+
+    /// Looks up a given method name on the given primitive type.
+    pub fn lookup_primitive_method(&self, typ: &Type, method_name: &str) -> Option<FuncId> {
+        get_type_method_key(typ)
+            .and_then(|key| self.primitive_methods.get(&(key, method_name.to_owned())).copied())
+    }
+}
+
+/// These are the primitive type variants that we support adding methods to
+#[derive(Copy, Clone, Hash, PartialEq, Eq)]
+enum TypeMethodKey {
+    /// Fields and integers share methods for ease of use. These methods may still
+    /// accept only fields or integers, it is just that their names may not clash.
+    FieldOrInt,
+    Array,
+    Bool,
+    String,
+    Unit,
+    Tuple,
+    Function,
```

Comment:
What are the implications of putting function here?

Reply:
It lets us add impls for functions if we really wanted them. In rust for example, you can impl your own traits on function types as well and some libraries make good use of this.

I don't know of any useful methods we could add off the top of my head but one example would be:
```rs
impl<T, U> fn(T) -> U {
    fn call(self, arg: T) -> U {
        self(arg)
    }
}

fn id<T>(x: T) -> T { x }

fn main() {
    constrain id.call(3) == id(3);
}
```

Reply:
Ah makes sense -- Thanks for clarifying ser